File: Data Structures & Algorithms/validate-parentheses/submission-5.py
class Solution:
    def isValid(self, s: str) -> bool:
        paren = { "]": '[', "}": "{", ")": "(" }
        stack = []
        for c in s:
            if c in paren and stack and stack[-1] == paren[c]:
                stack.pop()
            else:
                stack.append(c)

        return len(stack) == 0 

File: Data Structures & Algorithms/daily-temperatures/submission-5.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        if (temperatures.size() < 1) return std::vector<int>{};
        
        std::vector<int> result(temperatures.size(), 0);
        std::stack<std::pair<int, int>> st;

        for (size_t i = 0; i < temperatures.size(); i++) {
            while (!st.empty() && temperatures[i] > st.top().first) {
                auto [temp, ind] = st.top();
                st.pop();

                result[ind] = i - ind; 
            }
            st.push({temperatures[i], i});
        }
        return result;
    }
};
